### PR TITLE
wgpu: Implement blend modes

### PR DESCRIPTION
**Pull request by Dinnerbone**
All blend modes should, as far as I am aware and have been able to test, work correctly.

Performance isn't amazing - there's definitely areas to improve and if you stress it with hundreds of blends you'll notice it chug. I don't think it should be an impact on a "normal" swf though, and a big part of the performance hit is from the lack of cacheAsBitmap.

### CODE REVIEW

**Comment by angstsmurf** [comment]
This seems to completely fix #4464, making *The Queen of Snakes* fully playable.

**Comment by Dinnerbone** [comment]
Updated with (I believe) fully functional blend modes, and with latest code from master.

It *looks* like everything works, but the performance is bad so I'm keeping this a draft for now.

**Comment by angstsmurf** [comment]
Hm, one of the later updates seems to have caused #4464 to regress. The triangles everywhere are back, but completely black rather than the previously blue stretched eye graphic. When I tested this four days ago, they were properly invisible.

File: render/wgpu/shaders/blend.wgsl [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            if (src.g <= 0.5) { out.g = (2.0 * src.g * dst.g); } else { out.g = (1.0 - 2.0 * (1.0 - dst.g) * (1.0 - src.g)); }
+            if (src.b <= 0.5) { out.b = (2.0 * src.b * dst.b); } else { out.b = (1.0 - 2.0 * (1.0 - dst.b) * (1.0 - src.b)); }
+            return out;
+        }
+    }
+}
+
+@fragment
+fn main_fragment(in: VertexOutput) -> @location(0) vec4<f32> {
+    // dst is the parent pixel we're blending onto - it is either 0 or 1 alpha.
```
Comment (Herschel) [inline render/wgpu/shaders/blend.wgsl:69]: Is this true? E.g. you a blended object onto a transparent bitmap with <100% alpha objects -- but I haven't looked through what intermediate buffers things get drawn to!

File: render/wgpu/shaders/blend.wgsl [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    var dst: vec4<f32> = textureSample(parent_texture, texture_sampler, in.uv);
+    // src is the pixel that we want to apply - it may have any alpha.
+    var src: vec4<f32> = textureSample(current_texture, texture_sampler, in.uv);
+
+    if (src.a > 0.0) {
+        if (blend.mode == 6) { // Add
+            return vec4<f32>(src.rgb + dst.rgb, 1.0);
```
Comment (Herschel) [inline render/wgpu/shaders/blend.wgsl:76]: `return src.rgba + dst.rgba;` to handle alpha
Example:
[blend_add.zip](https://github.com/ruffle-rs/ruffle/files/10255203/blend_add.zip)

File: render/wgpu/shaders/blend.wgsl [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        } else if (blend.mode == 7) { // Subtract
+            return vec4<f32>(dst.rgb - src.rgb, 1.0);
```
Comment (Herschel) [inline render/wgpu/shaders/blend.wgsl:78]: `return vec4<f32>(dst.rgb - src.rgb, src.a + dst.a);`
Example:
[blend_subtract.zip](https://github.com/ruffle-rs/ruffle/files/10255206/blend_subtract.zip)

**Review by Herschel — COMMENTED** [review]
Thanks so much for your effort on this! It'll be very exciting to get this in.

Here are some quickie fixes for the additive modes that I saw. I can take a look at Erase etc. later. As I know your availability is limited, if you'd like, I can go ahead and push fixes to your branch.

**Comment by n0samu** [comment]
At first I got confused when testing because I was still seeing this message. I guess it should be removed, right?
https://github.com/ruffle-rs/ruffle/blob/7b32abeadfe7d42291407f68b074cb849c67a931/core/src/display_object.rs#L397

**Comment by n0samu** [comment]
Testing [The Last Stand: Union City](https://uploads.ungrounded.net/580000/580377_TheLastStand_UnionCity.swf) with `-g dx12`, I'm getting a crash while the game is loading:
```
[2022-12-19T07:54:27Z ERROR wgpu_hal::dx12::descriptor] Unable to allocate descriptors: RangeAllocationError { fragmented_free_length: 0 }
[2022-12-19T07:54:27Z ERROR wgpu::backend::direct] Handling wgpu errors as fatal by default
thread 'main' panicked at 'wgpu error: Validation Error

Caused by:
    In Device::create_bind_group
    not enough memory left

', C:\Users\nosamu\.cargo\registry\src\github.com-1ecc6299db9ec823\wgpu-0.14.2\src\backend\direct.rs:2403:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
No such issue occurs with the Vulkan renderer.

**Comment by Yuri-K7** [comment]
Seems like https://github.com/ruffle-rs/ruffle/issues/7820 also regressed... So there's at least a problem with darken blend mode...

**Comment by Dinnerbone** [comment]
Oh yep absolutely!

File: render/wgpu/shaders/blend.wgsl (hunk 1 above)
Comment (Dinnerbone) [reply render/wgpu/shaders/blend.wgsl:69]: You're right! Not sure what I was thinking :D

**Comment by Benman2785** [comment]
thank you - would be great to have for web-selfhosted ;)

**Comment by n0samu** [comment]
With this PR, the highlight effect works when hovering the mouse over the gems in [Bejeweled 2 (King.com version)](https://github.com/ruffle-rs/ruffle/files/10286637/bejeweled2.zip).

I also tried building Ruffle-selfhosted with wgpu-webgl to test this out on web. I built with dual-wasm enabled for best performance, but the performance running this Bejeweled game was still unacceptable in both Firefox and Chromium.

**Comment by Benman2785** [comment]
here some performance testing for web-selfhosted: 
POOR: [The last Stand 1](https://hoffmeister.li/flash-games/The_last_Stand.html)
FINE: [The last Stand 2](https://hoffmeister.li/flash-games/The_last_Stand_2.html)
FINE - but still bugged: [The last Stand 3](https://hoffmeister.li/flash-games/The_last_Stand_Union-City_wgpu.html)
the yellow screen is related to a resize bug - as the Win64 ruffle nightly also has this issue it is mostly because ruffle calculates the twips wrong // its a known bug for that game if resized (but can be fixed)

**Review by Aaron1011 — APPROVED** [review]

**Pull request merged**